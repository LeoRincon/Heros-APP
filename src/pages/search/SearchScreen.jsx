import React, { useMemo } from 'react';
import { useNavigate, useLocation } from 'react-router-dom';
import queryString from 'query-string';

import { useForm } from '../../hooks/useForm';
import { getHeroesByName } from '../../components/selectors/getHeroesByName';
import { renderHeroFound } from '../../components/selectors/renderHeroFound';

export const SearchScreen = () => {
  const { search } = useLocation();
  const { q: searchQuery = '' } = queryString.parse(search);

  const navigate = useNavigate();
  const [formValues, handleInputChange] = useForm({
    searchText: searchQuery,
  });
  const { searchText } = formValues;

  const heroesFiltered = useMemo(
    () => getHeroesByName(searchQuery),
    [searchQuery]
  );

  const handleSearch = (e) => {
    e.preventDefault();
    navigate(`?q=${searchText}`);
  };

  return (
    <section className='container'>
      <h2>SearchScreen</h2>
      <div className='row'>
        <div className='col-5'>
          <h4>Search Form</h4>
          <form onSubmit={handleSearch}>
            <input
              type='text'
              placeholder='Find for Hero...'
              className='form-control'
              autoComplete='off'
              name='searchText'
              value={searchText}
              onChange={handleInputChange}
            />
            <button
              type='submit'
              className='btn m-1 btn-block btn-outline-primary'
            >
              Search
            </button>
          </form>
        </div>
        <div className='col-7'></div>
        <h4>Result</h4>
        {searchQuery === '' && (
          <p className='alert alert-info'>Search a Hero</p>
        )}
        {searchQuery !== '' && heroesFiltered.length === 0 ? (
          <p className='alert alert-danger'>
            {`${searchQuery}: Heroes Found ingresa un nombre valido en el input`}
          </p>
        ) : (
          renderHeroFound(heroesFiltered)
        )}
      </div>
    </section>
  );
};
