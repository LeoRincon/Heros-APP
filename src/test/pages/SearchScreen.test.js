import { mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';

import { SearchScreen } from '../../pages/search/SearchScreen';

const mockNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockNavigate,
}));

describe('test in <SearchScreen /> when is called', () => {
  test('should be displayed correctly with dafault values', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search']}>
        <SearchScreen />
      </MemoryRouter>
    );
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('.alert-info').text().trim()).toBe('Search a Hero');
  });

  test('should show anyone Hero and the input with queryString value', () => {
    let anyOneHero = 'batman';
    const wrapper = mount(
      <MemoryRouter initialEntries={[`/search?q=${anyOneHero}`]}>
        <SearchScreen />
      </MemoryRouter>
    );
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('input').prop('value')).toBe(anyOneHero);
  });
  test('should show error if not find the hero', () => {
    let anyOneHero = 'anyOneHero';

    const wrapper = mount(
      <MemoryRouter initialEntries={[`/search?q=${anyOneHero}`]}>
        <SearchScreen />
      </MemoryRouter>
    );
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('.alert-danger').text().trim()).toBe(
      `${anyOneHero}: Heroes Found ingresa un nombre valido en el input`
    );
  });

  test('Should call the navigate to the new screen', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search']}>
        <SearchScreen />
      </MemoryRouter>
    );
    wrapper.find('input').simulate('change', {
      target: { name: 'searchText', value: 'batman' },
    });

    wrapper.find('form').prop('onSubmit')({
      preventDefault: () => {},
    });

    expect(mockNavigate).toHaveBeenCalledWith('?q=batman');
  });
});
