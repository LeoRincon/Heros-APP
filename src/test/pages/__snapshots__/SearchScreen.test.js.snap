// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test in <SearchScreen /> when is called should be displayed correctly with dafault values 1`] = `
<section
  className="container"
>
  <h2>
    SearchScreen
  </h2>
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find for Hero..."
          type="text"
          value=""
        />
        <button
          className="btn m-1 btn-block btn-outline-primary"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    />
    <h4>
      Result
    </h4>
    <p
      className="alert alert-info"
    >
      Search a Hero
    </p>
  </div>
</section>
`;

exports[`test in <SearchScreen /> when is called should show anyone Hero and the input with queryString value 1`] = `
<section
  className="container"
>
  <h2>
    SearchScreen
  </h2>
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find for Hero..."
          type="text"
          value="batman"
        />
        <button
          className="btn m-1 btn-block btn-outline-primary"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    />
    <h4>
      Result
    </h4>
    <div
      className="card ms-3 animate__animated animate__fadeIn"
      style={
        Object {
          "maxWidth": 540,
        }
      }
    >
      <div
        className="row no-gutters"
      >
        <div
          className="col-md-4"
        >
          <img
            alt="Batman"
            className="card-img"
            src="./assets/heroes/dc-batman.jpg"
          />
        </div>
        <div
          className="col-md-8"
        >
          <div
            className="card-body"
          >
            <h5
              className="card-title"
            >
              Batman
               (
              Bruce Wayne
              )
            </h5>
            <p
              className="card-text mb-5"
            >
               
              First Appearance:
              <small
                className="text-muted"
              >
                 
                Detective Comics #27
              </small>
            </p>
            <a
              className="btn btn-primary "
              href="/hero/dc-batman"
              onClick={[Function]}
            >
              Ver m√°s
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
</section>
`;

exports[`test in <SearchScreen /> when is called should show error if not find the hero 1`] = `
<section
  className="container"
>
  <h2>
    SearchScreen
  </h2>
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find for Hero..."
          type="text"
          value="anyOneHero"
        />
        <button
          className="btn m-1 btn-block btn-outline-primary"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    />
    <h4>
      Result
    </h4>
    <p
      className="alert alert-danger"
    >
      anyOneHero: Heroes Found ingresa un nombre valido en el input
    </p>
  </div>
</section>
`;
